<!DOCTYPE html>
<html lang="" xml:lang="">
  <head>
    <title>ggtranscript: Visualizing transcript structure and annotation using ggplot2</title>
    <meta charset="utf-8" />
    <meta name="author" content="David Zhang &amp; Emil Gustavsson" />
    <script src="libs/header-attrs/header-attrs.js"></script>
    <link href="libs/remark-css/default.css" rel="stylesheet" />
    <link href="libs/remark-css/metropolis.css" rel="stylesheet" />
    <link href="libs/remark-css/metropolis-fonts.css" rel="stylesheet" />
  </head>
  <body>
    <textarea id="source">
class: center, middle, inverse, title-slide

# ggtranscript: Visualizing transcript structure and annotation using ggplot2
### David Zhang &amp; Emil Gustavsson
### UCL
### 2022/02/04 (updated: 2022-03-17)

---


background-image: url("git_hooks.png")
background-size: 500px
background-position: 50% 80%



&lt;!-- change font of all slides via css --&gt;
&lt;!-- https://stackoverflow.com/questions/53481699/customize-font-size-for-all-the-slides-in-xaringan --&gt;

&lt;style type="text/css"&gt;
.remark-slide-content {
    font-size: 34px;
    padding: 0.75em 1em 0.75em 1em;
}
.remark-code, .remark-inline-code { 
    font-family: 'Source Code Pro', 'Lucida Console', Monaco, monospace;
    font-size: 30%;
}
&lt;/style&gt;

&lt;style type="text/css"&gt;
table {
  font-size: 14px;     
}

/* custom.css */
.left-code {
  color: #777;
  width: 38%;
  height: 92%;
  float: left;
}
.right-plot {
  width: 60%;
  float: right;
  padding-left: 1%;
}
&lt;/style&gt;

# Visualizing transcript structures and annotation

- The core components of transcript structures are exons and introns
- Exons can be broken down into CDS and UTRs
- Currently visualized through Ensembl/IGV/Gviz

---

background-image: url("https://raw.githubusercontent.com/dzhang32/ggtranscript/main/man/figures/ggtranscript_logo.png")
background-size: 300px
background-position: 50% 80%

# Introducing ggtranscript

- https://github.com/dzhang32/ggtranscript
- R package available through GitHub, soon to be Bioconductor
- ggplot2 extension, inherits functionality and flexibility

---

# Introducing ggtranscript

- Introduces 5 new geoms
- geom_range(), geom_half_range(), geom_intron(), geom_junction() and geom_junction_label_repel()
- And several helper functions
- Feedback is much appreciated! ðŸ˜„

---

# Required input data

- Data formats are widely used in genetic and transcriptomic analyses
- Requires a data.frame()


|Required aes() |Type                 |Description                                                       |Required by                      |
|:--------------|:--------------------|:-----------------------------------------------------------------|:--------------------------------|
|xstart         |integer              |Start position in base pairs                                      |All geoms                        |
|xend           |integer              |End position in base pairs                                        |All geoms                        |
|y              |charactor or factor  |The group used for the y axis, most often a transcript id or name |All geoms                        |
|label          |integer or charactor |Variable used to label junction curves                            |Only geom_junction_label_repel() |

---

# Plotting exons and introns

.left-code[

```r
sod1_exons &lt;- sod1_annotation %&gt;% dplyr::filter(type == "exon")

sod1_exons %&gt;%
    ggplot(aes(
        xstart = start,
        xend = end,
        y = transcript_name
    )) +
    geom_range(
        aes(fill = transcript_biotype)
    ) +
    geom_intron(
        data = to_intron(sod1_exons, "transcript_name"),
        aes(strand = strand)
    )
```
]

.right-plot[
&lt;img src="ggtranscript_files/figure-html/plot-exons-introns-out-1.png" width="100%" /&gt;
]

---

# Differentiating UTRs from the coding sequence

.left-code[

```r
sod1_exons_prot_cod &lt;- sod1_exons %&gt;%
    dplyr::filter(transcript_biotype == "protein_coding")
sod1_cds &lt;- sod1_annotation %&gt;% dplyr::filter(type == "CDS")

sod1_exons_prot_cod %&gt;%
    ggplot(aes(
        xstart = start,
        xend = end,
        y = transcript_name
    )) +
    geom_range(
        fill = "white",
        height = 0.25
    ) +
    geom_range(
        data = sod1_cds
    ) +
    geom_intron(
        data = to_intron(sod1_exons_prot_cod, "transcript_name"),
        aes(strand = strand),
        arrow.min.intron.length = 500,
    )
```
]

.right-plot[
&lt;img src="ggtranscript_files/figure-html/plot-cds-out-1.png" width="100%" /&gt;
]

???

`geom_range()` is designed to visualize range-based transcript annotation. This includes but is not limited to exons. For instance, for protein coding transcripts it can be useful to visually distinguish the coding sequence (CDS) of a transcript from it's UTRs.

---

# Plotting junctions

.left-code[

```r
sod1_201_exons &lt;- sod1_exons %&gt;% dplyr::filter(transcript_name == "SOD1-201")
sod1_201_cds &lt;- sod1_cds %&gt;% dplyr::filter(transcript_name == "SOD1-201")

sod1_junctions &lt;- sod1_junctions %&gt;% dplyr::mutate(transcript_name = "SOD1-201")

sod1_201_exons %&gt;%
  ggplot(aes(
    xstart = start,
    xend = end,
    y = transcript_name
  )) +
  geom_range(
    fill = "white", 
    height = 0.25
  ) +
  geom_range(
    data = sod1_201_cds
  ) + 
  geom_intron(
    data = to_intron(sod1_201_exons, "transcript_name")
  ) + 
  geom_junction(
    data = sod1_junctions,
    aes(size = mean_count),
    junction.y.max = 0.5
  ) +
  geom_junction_label_repel(
    data = sod1_junctions,
    aes(label = round(mean_count, 2)),
    junction.y.max = 0.5
  ) + 
  scale_size_continuous(range = c(0.1, 1))
```
]

.right-plot[
&lt;img src="ggtranscript_files/figure-html/plot-junctions-out-1.png" width="100%" /&gt;
]

---

# Inheriting ggplot2's functionality

.left-code[

```r
sod1_201_exons %&gt;%
  ggplot(aes(
    xstart = start,
    xend = end,
    y = transcript_name
  )) +
  geom_range(
    fill = "white", 
    height = 0.25
  ) +
  geom_range(
    data = sod1_201_cds
  ) + 
  geom_intron(
    data = to_intron(sod1_201_exons, "transcript_name")
  ) + 
  geom_junction(
    data = sod1_junctions,  
*   colour = "purple",
    aes(size = mean_count),
    junction.y.max = 0.5, 
    ncp = 30
  ) +
  scale_size_continuous(range = c(0.1, 1)) + 
* theme_bw()
```
]

.right-plot[
&lt;img src="ggtranscript_files/figure-html/plot-junctions-2-out-1.png" width="100%" /&gt;
]

---

# Visualizing transcript structure differences

.left-code[

```r
pknox1_exons &lt;- pknox1_annotation %&gt;% dplyr::filter(type == "exon")
pknox1_exons %&gt;%
    ggplot(aes(
        xstart = start,
        xend = end,
        y = transcript_name
    )) +
    geom_range(
        aes(fill = transcript_biotype)
    ) +
    geom_intron(
        data = to_intron(pknox1_exons, "transcript_name"),
        aes(strand = strand), 
        arrow.min.intron.length = 3500
    )
```
]

.right-plot[
&lt;img src="ggtranscript_files/figure-html/plot-diff-out-1.png" width="100%" /&gt;
]

---

# Improving transcript visualization with shorten_gaps()

.left-code[

```r
# extract exons
pknox1_exons &lt;- pknox1_annotation %&gt;% dplyr::filter(type == "exon")
*pknox1_rescaled &lt;- shorten_gaps(
  exons = pknox1_exons, 
  introns = to_intron(pknox1_exons, "transcript_name"), 
  group_var = "transcript_name"
)
# shorten_gaps() returns exons and introns all in one data.frame()
# let's split these for plotting 
pknox1_rescaled_exons &lt;- pknox1_rescaled %&gt;% dplyr::filter(type == "exon") 
pknox1_rescaled_introns &lt;- pknox1_rescaled %&gt;% dplyr::filter(type == "intron") 
pknox1_rescaled_exons %&gt;% 
    ggplot(aes(
        xstart = start,
        xend = end,
        y = transcript_name
    )) +
    geom_range(
        aes(fill = transcript_biotype)
    ) +
    geom_intron(
        data = pknox1_rescaled_introns,
        aes(strand = strand), 
        arrow.min.intron.length = 300
    )
```
]

.right-plot[
&lt;img src="ggtranscript_files/figure-html/plot-diff-2-out-1.png" width="100%" /&gt;
]

---

# Comparing two transcripts with geom_half_range()

.left-code[

```r
# extract the two transcripts to be compared
pknox1_rescaled_201_exons &lt;- pknox1_rescaled_exons %&gt;% 
  dplyr::filter(transcript_name == "PKNOX1-201")
pknox1_rescaled_203_exons &lt;- pknox1_rescaled_exons %&gt;% 
  dplyr::filter(transcript_name == "PKNOX1-203")
pknox1_rescaled_201_exons %&gt;%
    ggplot(aes(
        xstart = start,
        xend = end,
        y = "PKNOX1-201/203"
    )) +
    geom_half_range() +
    geom_intron(
        data = to_intron(pknox1_rescaled_201_exons, "transcript_name"), 
        arrow.min.intron.length = 300
    ) +
    geom_half_range(
        data = pknox1_rescaled_203_exons,
        range.orientation = "top", 
        fill = "purple"
    ) +
    geom_intron(
        data = to_intron(pknox1_rescaled_203_exons, "transcript_name"), 
        arrow.min.intron.length = 300
    )
```
]

.right-plot[
&lt;img src="ggtranscript_files/figure-html/plot-diff-3-out-1.png" width="100%" /&gt;
]

---

# Zooming using ggforce::facet_zoom()

.left-code[

```r
pknox1_rescaled_201_exons %&gt;%
    ggplot(aes(
        xstart = start,
        xend = end,
        y = "PKNOX1-201/203"
    )) +
    geom_half_range() +
    geom_intron(
        data = to_intron(pknox1_rescaled_201_exons, "transcript_name"), 
        arrow.min.intron.length = 300
    ) +
    geom_half_range(
        data = pknox1_rescaled_203_exons,
        range.orientation = "top", 
        fill = "purple"
    ) +
    geom_intron(
        data = to_intron(pknox1_rescaled_203_exons, "transcript_name"), 
        arrow.min.intron.length = 300
    ) + 
    ggforce::facet_zoom(xlim = c(0, 300))
```
]

.right-plot[
&lt;img src="ggtranscript_files/figure-html/plot-diff-5-out-1.png" width="100%" /&gt;
]


---

# Comparing to a single reference transcript with to_diff()

.left-code[

```r
mane &lt;- pknox1_rescaled_201_exons
not_mane &lt;- pknox1_rescaled_exons %&gt;% 
  dplyr::filter(transcript_name != "PKNOX1-201")

*pknox1_rescaled_diffs &lt;- to_diff(
    exons = not_mane,
    ref_exons = mane,
    group_var = "transcript_name"
)

pknox1_rescaled_exons %&gt;%
    ggplot(aes(
        xstart = start,
        xend = end,
        y = transcript_name
    )) +
    geom_range() +
    geom_intron(
        data = pknox1_rescaled_introns,
        arrow.min.intron.length = 300
    ) +
    geom_range(
        data = pknox1_rescaled_diffs,
        aes(fill = diff_type),
        alpha = 0.2
    )
```
]

.right-plot[
&lt;img src="ggtranscript_files/figure-html/plot-diff-4-out-1.png" width="100%" /&gt;
]

---

# Conclusions

- ggtranscript makes it easy to visualize and compare transcript structures 
- Lots of functionality as a ggplot2 extension
- Faster and more flexible than existing tools

---

class: center, middle, inverse

# Thank you for listening! ðŸ˜„

    </textarea>
<style data-target="print-only">@media screen {.remark-slide-container{display:block;}.remark-slide-scaler{box-shadow:none;}}</style>
<script src="https://remarkjs.com/downloads/remark-latest.min.js"></script>
<script>var slideshow = remark.create({
"highlightStyle": "github",
"highlightLines": true,
"countIncrementalSlides": false,
"ratio": "16:9"
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {
  window.dispatchEvent(new Event('resize'));
});
(function(d) {
  var s = d.createElement("style"), r = d.querySelector(".remark-slide-scaler");
  if (!r) return;
  s.type = "text/css"; s.innerHTML = "@page {size: " + r.style.width + " " + r.style.height +"; }";
  d.head.appendChild(s);
})(document);

(function(d) {
  var el = d.getElementsByClassName("remark-slides-area");
  if (!el) return;
  var slide, slides = slideshow.getSlides(), els = el[0].children;
  for (var i = 1; i < slides.length; i++) {
    slide = slides[i];
    if (slide.properties.continued === "true" || slide.properties.count === "false") {
      els[i - 1].className += ' has-continuation';
    }
  }
  var s = d.createElement("style");
  s.type = "text/css"; s.innerHTML = "@media print { .has-continuation { display: none; } }";
  d.head.appendChild(s);
})(document);
// delete the temporary CSS (for displaying all slides initially) when the user
// starts to view slides
(function() {
  var deleted = false;
  slideshow.on('beforeShowSlide', function(slide) {
    if (deleted) return;
    var sheets = document.styleSheets, node;
    for (var i = 0; i < sheets.length; i++) {
      node = sheets[i].ownerNode;
      if (node.dataset["target"] !== "print-only") continue;
      node.parentNode.removeChild(node);
    }
    deleted = true;
  });
})();
(function() {
  "use strict"
  // Replace <script> tags in slides area to make them executable
  var scripts = document.querySelectorAll(
    '.remark-slides-area .remark-slide-container script'
  );
  if (!scripts.length) return;
  for (var i = 0; i < scripts.length; i++) {
    var s = document.createElement('script');
    var code = document.createTextNode(scripts[i].textContent);
    s.appendChild(code);
    var scriptAttrs = scripts[i].attributes;
    for (var j = 0; j < scriptAttrs.length; j++) {
      s.setAttribute(scriptAttrs[j].name, scriptAttrs[j].value);
    }
    scripts[i].parentElement.replaceChild(s, scripts[i]);
  }
})();
(function() {
  var links = document.getElementsByTagName('a');
  for (var i = 0; i < links.length; i++) {
    if (/^(https?:)?\/\//.test(links[i].getAttribute('href'))) {
      links[i].target = '_blank';
    }
  }
})();
// adds .remark-code-has-line-highlighted class to <pre> parent elements
// of code chunks containing highlighted lines with class .remark-code-line-highlighted
(function(d) {
  const hlines = d.querySelectorAll('.remark-code-line-highlighted');
  const preParents = [];
  const findPreParent = function(line, p = 0) {
    if (p > 1) return null; // traverse up no further than grandparent
    const el = line.parentElement;
    return el.tagName === "PRE" ? el : findPreParent(el, ++p);
  };

  for (let line of hlines) {
    let pre = findPreParent(line);
    if (pre && !preParents.includes(pre)) preParents.push(pre);
  }
  preParents.forEach(p => p.classList.add("remark-code-has-line-highlighted"));
})(document);</script>

<script>
slideshow._releaseMath = function(el) {
  var i, text, code, codes = el.getElementsByTagName('code');
  for (i = 0; i < codes.length;) {
    code = codes[i];
    if (code.parentNode.tagName !== 'PRE' && code.childElementCount === 0) {
      text = code.textContent;
      if (/^\\\((.|\s)+\\\)$/.test(text) || /^\\\[(.|\s)+\\\]$/.test(text) ||
          /^\$\$(.|\s)+\$\$$/.test(text) ||
          /^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)) {
        code.outerHTML = code.innerHTML;  // remove <code></code>
        continue;
      }
    }
    i++;
  }
};
slideshow._releaseMath(document);
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
