---
title: "ggtranscript: Visualizing transcript structure and annotation using ggplot2"
author: "David Zhang"
institute: "UCL"
date: "2022/02/04 (updated: `r Sys.Date()`)"
output:
  xaringan::moon_reader:
    css: [default, metropolis, metropolis-fonts]
    lib_dir: libs
    nature:
      highlightStyle: github
      highlightLines: true
      countIncrementalSlides: false
      ratio: 16:9
---

background-image: url("git_hooks.png")
background-size: 500px
background-position: 50% 80%

```{r setup, include=FALSE}
library(tidyverse)
library(ggtranscript)

options(htmltools.dir.version = FALSE)
knitr::opts_chunk$set(fig.dim = c(10, 6), fig.retina=2, out.width="100%")

hook_source <- knitr::knit_hooks$get('source')
knitr::knit_hooks$set(source = function(x, options) {
  x <- stringr::str_replace(x, "^[[:blank:]]?([^*].+?)[[:blank:]]*#<<[[:blank:]]*$", "*\\1")
  hook_source(x, options)
})

```

<!-- change font of all slides via css -->
<!-- https://stackoverflow.com/questions/53481699/customize-font-size-for-all-the-slides-in-xaringan -->

<style type="text/css">
.remark-slide-content {
    font-size: 34px;
    padding: 0.75em 1em 0.75em 1em;
}
.remark-code, .remark-inline-code { 
    font-family: 'Source Code Pro', 'Lucida Console', Monaco, monospace;
    font-size: 30%;
}
</style>

```{css, echo = F}
table {
  font-size: 14px;     
}

/* custom.css */
.left-code {
  color: #777;
  width: 38%;
  height: 92%;
  float: left;
}
.right-plot {
  width: 60%;
  float: right;
  padding-left: 1%;
}
```

# Visualizing transcript structures and annotation

- The core components of transcript structures are exons and introns
- Exons can be broken down into CDS and UTRs
- Currently visualized through Ensembl/IGV/Gviz

---

background-image: url("https://raw.githubusercontent.com/dzhang32/ggtranscript/main/man/figures/ggtranscript_logo.png")
background-size: 300px
background-position: 50% 80%

# Introducing ggtranscript

- https://github.com/dzhang32/ggtranscript
- R package available through GitHub, soon to be Bioconductor
- ggplot2 extension, inherits functionality and flexibility

---

# Introducing ggtranscript

- Introduces 5 new geoms
- geom_range(), geom_half_range(), geom_intron(), geom_junction() and geom_junction_label_repel()
- And several helper functions
- Feedback is much appreciated! `r emo::ji("smile")`

---

# Required input data

- Data formats are widely used in genetic and transcriptomic analyses
- Requires a data.frame()

```{r, echo = FALSE}

dplyr::tribble(
    ~`Required aes()`, ~Type, ~Description, ~`Required by`,
    #-------|-----|----|--------
    "xstart", "integer", "Start position in base pairs", "All geoms", 
    "xend", "integer", "End position in base pairs", "All geoms", 
    "y", "charactor or factor", "The group used for the y axis, most often a transcript id or name ", "All geoms",
    "label", "integer or charactor", "Variable used to label junction curves", "Only geom_junction_label_repel()",
) %>% 
  knitr::kable()

```

---

# Plotting exons and introns

.left-code[
```{r plot-exons-introns, eval=FALSE}
sod1_exons <- sod1_annotation %>% dplyr::filter(type == "exon")

sod1_exons %>%
    ggplot(aes(
        xstart = start,
        xend = end,
        y = transcript_name
    )) +
    geom_range(
        aes(fill = transcript_biotype)
    ) +
    geom_intron(
        data = to_intron(sod1_exons, "transcript_name"),
        aes(strand = strand)
    )
```
]

.right-plot[
```{r plot-exons-introns-out, ref.label="plot-exons-introns", echo=FALSE}
```
]

---

# Differentiating UTRs from the coding sequence

.left-code[
```{r plot-cds, eval=FALSE}

sod1_exons_prot_cod <- sod1_exons %>%
    dplyr::filter(transcript_biotype == "protein_coding")
sod1_cds <- sod1_annotation %>% dplyr::filter(type == "CDS")

sod1_exons_prot_cod %>%
    ggplot(aes(
        xstart = start,
        xend = end,
        y = transcript_name
    )) +
    geom_range(
        fill = "white",
        height = 0.25
    ) +
    geom_range(
        data = sod1_cds
    ) +
    geom_intron(
        data = to_intron(sod1_exons_prot_cod, "transcript_name"),
        aes(strand = strand),
        arrow.min.intron.length = 500,
    )
```
]

.right-plot[
```{r plot-cds-out, ref.label="plot-cds", echo=FALSE}
```
]

???

`geom_range()` is designed to visualize range-based transcript annotation. This includes but is not limited to exons. For instance, for protein coding transcripts it can be useful to visually distinguish the coding sequence (CDS) of a transcript from it's UTRs.

---

# Plotting junctions

.left-code[
```{r plot-junctions, eval=FALSE}


sod1_201_exons <- sod1_exons %>% dplyr::filter(transcript_name == "SOD1-201")
sod1_201_cds <- sod1_cds %>% dplyr::filter(transcript_name == "SOD1-201")

sod1_junctions <- sod1_junctions %>% dplyr::mutate(transcript_name = "SOD1-201")

sod1_201_exons %>%
  ggplot(aes(
    xstart = start,
    xend = end,
    y = transcript_name
  )) +
  geom_range(
    fill = "white", 
    height = 0.25
  ) +
  geom_range(
    data = sod1_201_cds
  ) + 
  geom_intron(
    data = to_intron(sod1_201_exons, "transcript_name")
  ) + 
  geom_junction(
    data = sod1_junctions,
    aes(size = mean_count),
    junction.y.max = 0.5
  ) +
  geom_junction_label_repel(
    data = sod1_junctions,
    aes(label = round(mean_count, 2)),
    junction.y.max = 0.5
  ) + 
  scale_size_continuous(range = c(0.1, 1))

```
]

.right-plot[
```{r plot-junctions-out, ref.label="plot-junctions", echo=FALSE}
```
]

---

# Inheriting ggplot2's functionality

.left-code[
```{r plot-junctions-2, eval=FALSE}

sod1_201_exons %>%
  ggplot(aes(
    xstart = start,
    xend = end,
    y = transcript_name
  )) +
  geom_range(
    fill = "white", 
    height = 0.25
  ) +
  geom_range(
    data = sod1_201_cds
  ) + 
  geom_intron(
    data = to_intron(sod1_201_exons, "transcript_name")
  ) + 
  geom_junction(
    data = sod1_junctions,  
    colour = "purple", #<<
    aes(size = mean_count),
    junction.y.max = 0.5, 
    ncp = 30
  ) +
  scale_size_continuous(range = c(0.1, 1)) + 
  theme_bw() #<<

```
]

.right-plot[
```{r plot-junctions-2-out, ref.label="plot-junctions-2", echo=FALSE}
```
]

---

# Visualizing transcript structure differences

.left-code[
```{r plot-diff, eval=FALSE}

pknox1_exons <- pknox1_annotation %>% dplyr::filter(type == "exon")
pknox1_exons %>%
    ggplot(aes(
        xstart = start,
        xend = end,
        y = transcript_name
    )) +
    geom_range(
        aes(fill = transcript_biotype)
    ) +
    geom_intron(
        data = to_intron(pknox1_exons, "transcript_name"),
        aes(strand = strand), 
        arrow.min.intron.length = 3500
    )

```
]

.right-plot[
```{r plot-diff-out, ref.label="plot-diff", echo=FALSE}
```
]

---

# Improving transcript visualization with shorten_gaps()

.left-code[
```{r plot-diff-2, eval=FALSE}

# extract exons
pknox1_exons <- pknox1_annotation %>% dplyr::filter(type == "exon")
pknox1_rescaled <- shorten_gaps( #<<
  exons = pknox1_exons, 
  introns = to_intron(pknox1_exons, "transcript_name"), 
  group_var = "transcript_name"
)
# shorten_gaps() returns exons and introns all in one data.frame()
# let's split these for plotting 
pknox1_rescaled_exons <- pknox1_rescaled %>% dplyr::filter(type == "exon") 
pknox1_rescaled_introns <- pknox1_rescaled %>% dplyr::filter(type == "intron") 
pknox1_rescaled_exons %>% 
    ggplot(aes(
        xstart = start,
        xend = end,
        y = transcript_name
    )) +
    geom_range(
        aes(fill = transcript_biotype)
    ) +
    geom_intron(
        data = pknox1_rescaled_introns,
        aes(strand = strand), 
        arrow.min.intron.length = 300
    )

```
]

.right-plot[
```{r plot-diff-2-out, ref.label="plot-diff-2", echo=FALSE}
```
]

---

# Comparing two transcripts with geom_half_range()

.left-code[
```{r plot-diff-3, eval=FALSE}

# extract the two transcripts to be compared
pknox1_rescaled_201_exons <- pknox1_rescaled_exons %>% 
  dplyr::filter(transcript_name == "PKNOX1-201")
pknox1_rescaled_203_exons <- pknox1_rescaled_exons %>% 
  dplyr::filter(transcript_name == "PKNOX1-203")
pknox1_rescaled_201_exons %>%
    ggplot(aes(
        xstart = start,
        xend = end,
        y = "PKNOX1-201/203"
    )) +
    geom_half_range() +
    geom_intron(
        data = to_intron(pknox1_rescaled_201_exons, "transcript_name"), 
        arrow.min.intron.length = 300
    ) +
    geom_half_range(
        data = pknox1_rescaled_203_exons,
        range.orientation = "top", 
        fill = "purple"
    ) +
    geom_intron(
        data = to_intron(pknox1_rescaled_203_exons, "transcript_name"), 
        arrow.min.intron.length = 300
    )

```
]

.right-plot[
```{r plot-diff-3-out, ref.label="plot-diff-3", echo=FALSE}
```
]

---

# Zooming using ggforce::facet_zoom()

.left-code[
```{r plot-diff-5, eval=FALSE}

pknox1_rescaled_201_exons %>%
    ggplot(aes(
        xstart = start,
        xend = end,
        y = "PKNOX1-201/203"
    )) +
    geom_half_range() +
    geom_intron(
        data = to_intron(pknox1_rescaled_201_exons, "transcript_name"), 
        arrow.min.intron.length = 300
    ) +
    geom_half_range(
        data = pknox1_rescaled_203_exons,
        range.orientation = "top", 
        fill = "purple"
    ) +
    geom_intron(
        data = to_intron(pknox1_rescaled_203_exons, "transcript_name"), 
        arrow.min.intron.length = 300
    ) + 
    ggforce::facet_zoom(xlim = c(0, 300))

```
]

.right-plot[
```{r plot-diff-5-out, ref.label="plot-diff-5", echo=FALSE}
```
]


---

# Comparing to a single reference transcript with to_diff()

.left-code[
```{r plot-diff-4, eval=FALSE}

mane <- pknox1_rescaled_201_exons
not_mane <- pknox1_rescaled_exons %>% 
  dplyr::filter(transcript_name != "PKNOX1-201")

pknox1_rescaled_diffs <- to_diff( #<<
    exons = not_mane,
    ref_exons = mane,
    group_var = "transcript_name"
)

pknox1_rescaled_exons %>%
    ggplot(aes(
        xstart = start,
        xend = end,
        y = transcript_name
    )) +
    geom_range() +
    geom_intron(
        data = pknox1_rescaled_introns,
        arrow.min.intron.length = 300
    ) +
    geom_range(
        data = pknox1_rescaled_diffs,
        aes(fill = diff_type),
        alpha = 0.2
    )

```
]

.right-plot[
```{r plot-diff-4-out, ref.label="plot-diff-4", echo=FALSE}
```
]

---

# Conclusions

- ggtranscript makes it easy to visualize and compare transcript structures 
- Lots of functionality as a ggplot2 extension
- Faster and more flexible than existing tools

---

class: center, middle, inverse

# Thank you for listening! `r emo::ji("smile")`

